### ISSUE TITLE
Multi-gpu pre-training

### ISSUE THREAD

**Issue by mtran14** [YuanGongND/cav-mae#6]
Dear Yuan, thank you for releasing the source code for your great project. It looks like the current code only supports pre-training on a single GPU (e.g., missing DDP data sampler) while your paper indicates that the released models were pre-trained on 4 GPUs. 

If possible, could you please update the code to support the multi-gpu pre-training?

Thank you.

**Comment by YuanGongND** [comment]
hi there, 

Could you elaborate this? 

The code is the actual code we used to train on 4 GPUs (of a single node). We used `torch.nn.DataParallel`

https://github.com/YuanGongND/cav-mae/blob/6cb02fe785f1f8cd3529376074652db08733c674/src/traintest_cavmae.py#L43-L44

Just because we used 4 GPUs, the losses on each GPU are added, and if you run on single GPU, the result will be slightly different. 

https://github.com/YuanGongND/cav-mae/blob/6cb02fe785f1f8cd3529376074652db08733c674/src/traintest_cavmae.py#L93-L94

-Yuan

**Comment by mtran14** [comment]
Thank you for the quick reply and pointing out the part using multi-GPUs. I mistakenly assumed that you are using the`torch.nn.parallel.DistributedDataParallel` module, which requires more steps. Sorry for the inconvenience caused.

**Comment by YuanGongND** [comment]
You actually made a great point. `torch.nn.DataParallel` is the easiest but the least effcient way for parallelization. `DistributedDataParallel` and more advanced methods (FSDP) are (much) more efficient. Given the fact that CAV-MAE pretraining is sensitive to the batch size, it would be nice to explore the best way to parallelize the model training.

Having that said, due to the contrastive loss, gradient accumulation won't be a feasible solution with a trivial solution.

-Yuan

**Comment by YuanGongND** [comment]
The code in this repo honestly reflects what we did in the paper. I did polish the code before the release to make it more clear, but as you can see in https://github.com/YuanGongND/cav-mae/tree/master/egs/audioset/training_logs, the results are almost exactly same with (or better than) that stated in the paper.
